control-plane/agents/src/bin/core/volume/operations.rs

```diff
@@ -1054,3 +1065,56 @@ impl CreateVolumeExeVal for CreateVolumeSource<'_> {
         }
     }
 }
+
+// spread_label_is_same checks if the node labels is same as that of all existing replicas.
+fn spread_label_is_same(
+    replica: &Replica,
+    replica_candidate: &CreateReplica,
+    context: &Context,
+) -> Result<bool, SvcError> {
+    let Some(ref topology) = context.volume.as_ref().topology else {
+        return Ok(false);
+    };
+
+    let Some(node_topology) = topology.node.as_ref() else {
+        return Ok(false);
+    };
+    let node_exclusion_label_match = match node_topology {
+        NodeTopology::Explicit(_) => true,
+        NodeTopology::Labelled(labelled_topology) => {
+            // If the NodeSpreadTopologyKey is present in the Volume Node Topology, then
+            // get the node details from replica candidate and
+            // check if the node labels is same as that of all existing replicas.
+            // If its same then don't consider the pool for replica creation
+            let mut exclusion_label_match = false;
+            if !labelled_topology.exclusion.is_empty() {
+                let candidate_node_label = context
+                    .registry
+                    .specs()
+                    .node(&replica_candidate.node)?
+                    .labels()
+                    .clone();
+
+                let replica_node_label = context
+                    .registry
+                    .specs()
+                    .node(&replica.node)?
+                    .labels()
+                    .clone();
+
+                for key in labelled_topology.exclusion.keys() {
+                    if let (Some(candidate_value), Some(replica_value)) =
+                        (candidate_node_label.get(key), replica_node_label.get(key))
+                    {
+                        if candidate_value == replica_value {
+                            exclusion_label_match = true;
```

Comment:
Why do we break out after a single match? What if the other keys don't match?

Reply:
In spread if any on the key matches , we must skip that node, thus we break even if a single key matches.

Reply:
Ah got it, btw we can just break out with return Ok(false), we don't need this variable